src/java/org/apache/cassandra/locator/DynamicEndpointSnitch.java

```diff
@@ -21,133 +21,505 @@
+        {
+            if (entry.getKey().equals(FBUtilities.getBroadcastAddressAndPort()))
+                continue;
+
+            AnnotatedMeasurement measurement = entry.getValue();
+            long lastMeasure = measurement.millisSinceLastMeasure.addAndGet(intervalMillis);
+            long lastRequest = measurement.millisSinceLastRequest.addAndGet(intervalMillis);
+
+            if (lastMeasure >= MIN_PROBE_INTERVAL_MS && lastRequest < MAX_PROBE_INTERVAL_MS)
+            {
+                if (measurement.nextProbeDelayMillis == 0)
+                {
+                    measurement.nextProbeDelayMillis = intervalMillis;
+                }
+                else if (measurement.probeFuture != null && measurement.probeFuture.isDone())
+                {
+                    measurement.nextProbeDelayMillis = Math.min(MAX_PROBE_INTERVAL_MS, measurement.nextProbeDelayMillis * 2);
+                }
+            }
+            else
+            {
+                measurement.nextProbeDelayMillis = 0;
+            }
+        }
     }
 
-    public void close()
+    @VisibleForTesting
+    void scheduleProbes(Map<InetAddressAndPort, AnnotatedMeasurement> samples)
     {
-        updateSchedular.cancel(false);
-        resetSchedular.cancel(false);
+        for (Map.Entry<InetAddressAndPort, AnnotatedMeasurement> entry: samples.entrySet())
+        {
+            AnnotatedMeasurement measurement = entry.getValue();
+            if (measurement.millisSinceLastRequest.get() > MAX_PROBE_INTERVAL_MS &&
+                !Gossiper.instance.isAlive(entry.getKey()))
+            {
+                samples.remove(entry.getKey());
+                continue;
+            }
 
-        MBeanWrapper.instance.unregisterMBean(mbeanName);
+            long delay = measurement.nextProbeDelayMillis;
+            if (delay > 0 && (measurement.probeFuture == null || measurement.probeFuture.isDone()))
+            {
+                logger.trace("Scheduled latency probe against {} in {}ms", entry.getKey(), delay);
+                measurement.probeFuture = latencyProbeExecutor.schedule(() -> sendPingMessageToPeer(entry.getKey()),
+                                                                        delay, TimeUnit.MILLISECONDS);
+            }
+        }
+    }
+
+    /**
+     * Method that actually sends latency probes as PING messages. This is the only function in this class
+     * that operates on the latencyProbeExecutor thread and it records the maximum latency between a small and large
+     * message channel ping.
+     */
+    private void sendPingMessageToPeer(InetAddressAndPort to)
+    {
+        // This method may have been scheduled (a long time) before it executes, so have to do
+        // some quick sanity checks before sending a message to this host
+        if (!StorageService.instance.isGossipActive() || !Gossiper.instance.isAlive(to))
+            return;
+
+        probeRateLimiter.acquire(dynamicSampleUpdateInterval);
```

Comment:
Reminder to fix the bug where this always does one a second.

Reply:
Ack, fixed.